File: src/problemset/binarySearch/SearchInsertPosition.java
package problemset.binarySearch;

class SearchInsertPosition {
    public int searchInsert(int[] nums, int target) {
        for (int i = 0; i < nums.length; i++) {
            if (target <= nums[i]) {
                return i;
            }
        }
        return nums.length;
    }
    public int binarySearchInsert(int[]  nums, int target) {
        /*
          烂尾方法
         */
        int left = 0, right = nums.length - 1;
        if (target <= nums[0])
            return 0;
        if (right == 0)
            return nums[0] > target ? 0 : 1;
        if (right == 1) {
            if (target >= nums[0] && target <= nums[1])
                return 1;
            else if (target <= nums[0])
                return 0;
            else
                return 2;
        }

        while (left <= right) {
            int mid = (left + right) / 2;
            if (nums[mid - 1] < target && target <= nums[mid]) 
                return mid;
            else if (nums[mid - 1] >= target)
                right = mid - 1;
            else
                left = mid + 1;
        }
        return nums.length;
    }

    public static void main(String args[]) {
        SearchInsertPosition solution = new SearchInsertPosition();
        System.out.println(solution.binarySearchInsert(new int[]{1,2,3,4,5,10}, 2));
    }
}

File: src/problemset/binarySearch/SearchInsertPosition.go
package binarySearch

// import "fmt"
// problem: 35
// url: https://leetcode-cn.com/problems/search-insert-position/
// Time: O(logn); 4ms, 90.19%
// Space: O(1); 3.1MB, 33.33%
func searchInsert(nums []int, target int) int {
	left, right := 0, len(nums)
	if target <= nums[0] {
		return 0
	}
	if right == 1 {
		return 1
	}

	for left < right {
		mid := left + (right - left) >> 1
		if nums[mid] >= target && nums[mid-1] < target {
			return mid
		} else if nums[mid] > target {
			right = mid
		} else {
			// nums[mid] < target
			left = mid + 1
		}
	}
	// fmt.Printf("(%d, %d)", left, right)
	return right
}

// func main() {
// 	fmt.Println(searchInsert([]int{1, 3, 5, 6}, 5))
// 	fmt.Println(searchInsert([]int{1, 3, 5, 6}, 2))
// 	fmt.Println(searchInsert([]int{1, 3, 5, 6}, 7))
// 	fmt.Println(searchInsert([]int{1, 3, 5, 6}, 0))
// 	fmt.Println(searchInsert([]int{1, 2, 3, 4, 5, 10}, 2))
// }
